### PR TITLE
feat(components): [el-message] config-provider message max attr

### LINKED ISSUE #4868: [Feature Request] el-message max attribute
### Existing Component
是

### Component Name
el-message

### Description
Add Max attribute for El message and can be set in El Config Provider

### PR DESCRIPTION
**Pull request by btea**
Please make sure these boxes are checked before submitting your PR, thank you!

- [ ] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [ ] Make sure you are merging your commits to `dev` branch.
- [ ] Add some descriptions and refer to relative issues for your PR.

feat #4868

### CODE REVIEW

File: packages/components/config-provider/__tests__/config-provider.spec.ts [hunk 1]
```diff
@@ -120,4 +120,38 @@ describe('config-provider', () => {
...
+        `,
+      })
+      await nextTick()
+      wrapper.find('.el-button').trigger('click')
+      wrapper.find('.el-button').trigger('click')
+      wrapper.find('.el-button').trigger('click')
+      wrapper.find('.el-button').trigger('click')
+      expect(document.querySelectorAll('.el-message').length).toBe(4)
+    })
+  })
```
Comment (JeremyWuuuuu) [inline packages/components/config-provider/__tests__/config-provider.spec.ts:156]: **Questions:** when you have the limitation set to max messages to `3`, why do you assert that there are 4 messages that shows up? I think this test case is completely wrong.

**Suggestion:** you should add `await` keyword before `.trigger([event])` statement.

File: packages/components/config-provider/src/config-provider.ts [hunk 2]
```diff
@@ -30,6 +37,7 @@ export default defineComponent({
 
   setup(props, { slots }) {
     provideLocale()
+    Object.assign(messageConfig, props.message || {})
```
Comment (JeremyWuuuuu) [inline packages/components/config-provider/src/config-provider.ts:40]: I think you should make sure that this value is reactive, for example:

```ts
watch(() => props.message, (val) => {
  Object.assign(messageConfig, val ?? {})
}, { immediate: true })
```

Comment (JeremyWuuuuu) [reply packages/components/config-provider/src/config-provider.ts:40]: Then you should add test against updating the message maximum value.

File: packages/components/message/src/message-method.ts [hunk 3]
```diff
@@ -16,6 +17,9 @@ let seed = 1
 
 const message: MessageFn & Partial<Message> = function (options = {}) {
   if (!isClient) return { close: () => undefined }
+  if (messageConfig.max && instances.length >= messageConfig.max) {
```
Comment (JeremyWuuuuu) [inline packages/components/message/src/message-method.ts:20]: We want to make sure that the value passed to `messageConfig.max` is a number.

Do:

```ts
isNumber(messageConfig.max) && instances.length >= messageConfig.max
```

**Review by jw-foss — CHANGES_REQUESTED** [review]
Left a couple comments for you to check, thanks.

File: packages/components/config-provider/__tests__/config-provider.spec.ts (hunk 1 above)
Comment (btea) [reply packages/components/config-provider/__tests__/config-provider.spec.ts:156]: > **Questions:** when you have the limitation set to max messages to `3`, why do you assert that there are 4 messages that shows up? I think this test case is completely wrong.

I'm sorry, this should be some conflicts caused by another branch merged with dev. I copied it directly and forgot to modify it. 🤣 
> **Suggestion:** you should add `await` keyword before `.trigger([event])` statement.

Maybe I just need to add an `await nextTick()` after all `.trigger([event])`.

File: packages/components/config-provider/src/config-provider.ts (hunk 2 above)
Comment (btea) [reply packages/components/config-provider/src/config-provider.ts:40]: You are right.

File: packages/components/message/src/message-method.ts (hunk 3 above)
Comment (btea) [reply packages/components/message/src/message-method.ts]: Done.

**Comment by emojiiii** [comment]
I think we should discuss how to deal(`discard/queue`) with redundant messages if the maximum is exceeded @btea @JeremyWuuuuu

**Comment by btea** [comment]
IMO, redundant message should be discarded. If multiple messages are placed in the queue and pop up one by one, the user may switch to other pages, but the code logic is still popping up messages one after another, which will cause confusion.

**Comment by emojiiii** [comment]
> IMO, redundant message should be discarded. If multiple messages are placed in the queue and pop up one by one, the user may switch to other pages, but the code logic is still popping up messages one after another, which will cause confusion.

But if they are discarded, some important information may be lost?

**Comment by btea** [comment]
> > IMO, redundant message should be discarded. If multiple messages are placed in the queue and pop up one by one, the user may switch to other pages, but the code logic is still popping up messages one after another, which will cause confusion.
> 
> But if they are discarded, some important information may be lost?

I think it depends on the user. If the user thinks the information is important, he can not set the max, or he can set the max a little larger.

**Comment by emojiiii** [comment]
can you test when you use two `config-provide` and different `max` ? @btea

**Comment by btea** [comment]
Generally, only one `config-provider` will be used. If there are more than one, the latter will overwrite the former, and the other configurations are the same.

**Comment by emojiiii** [comment]
You're right. If you can add another `unit test` for this, it will be very robust

**Comment by btea** [comment]
Do you mean to add multiple `config-provider` and then set different `max` values?

**Comment by emojiiii** [comment]
> Do you mean to add multiple `config-provider` and then set different `max` values?

yes

**Comment by btea** [comment]
Maybe it is better to modify the value of config directly like in the test code?

**Comment by emojiiii** [comment]
> Maybe it is better to modify the value of config directly like in the test code?

I'm just worried about whether multiple `config-provider` can merge configuration items as expected

**Comment by btea** [comment]
Okay, I will add unit test about this.

**Comment by JeremyWuuuuu** [comment]
> > Maybe it is better to modify the value of config directly like in the test code?
> 
> I'm just worried about whether multiple `config-provider` can merge configuration items as expected

IMHO, multiple provider is made by the developer, so that it is at their choice to make the call, so that we cannot handle this because they did what they did, that was completely on their own, so that we can just ignore it. Because the specialty about these global items (as designed this way) they should be able to accept that. Or maybe later we can try to pass the context to the call so that we can get the app context.

**Review by jw-foss — APPROVED** [review]
LGTM
